### PR TITLE
DOC: update str.cat example

### PR DESCRIPTION
**Pull request by h-vetinari**
There have been a couple of changes recently due to the whole docstring validation effort (e.g. #22838), which got rid of some warnings emitted by the sample code.

As it stands, the code samples don't really reflect the running text anymore, and with the changes from #22264, it makes sense to update the running text as well.

### CODE REVIEW

**Comment by h-vetinari** [comment]
Ok, the build didn't fail - it's just a flaky hypothesis failure.

File: doc/source/text.rst [hunk 1]
```diff
@@ -306,21 +306,24 @@ The same alignment can be used when ``others`` is a ``DataFrame``:
...
+Several items can be combined a list-like container (including iterators, ``dict``-views, etc.), which may contain ``Series``, ``Index`` and ``np.ndarray``.
+Note that ``Index`` will align as well, so we change the indexes of ``s`` and ``u`` to strings for the purpose of this example:
 
 .. ipython:: python
 
-    s
-    u
-    s.str.cat([u.values,
-               u.index.astype(str).values], na_rep='-')
+    s2 = s.set_axis(['a', 'b', 'c', 'd'], inplace=False)
```
Comment (jreback) [inline doc/source/text.rst]: this is not obvious at all, just directly construct the Series will be much more clear

File: doc/source/text.rst [hunk 2]
```diff
@@ -306,21 +306,24 @@ The same alignment can be used when ``others`` is a ``DataFrame``:
...
+    s2
+    u2 = u.set_axis(['b', 'd', 'a', 'c'], inplace=False)
```
Comment (jreback) [inline doc/source/text.rst]: same here

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/text.rst [hunk 3]
```diff
@@ -306,21 +306,24 @@ The same alignment can be used when ``others`` is a ``DataFrame``:
...
 
-All elements must match in length to the calling ``Series`` (or ``Index``), except those having an index if ``join`` is not None:
+All ``np.ndarrays`` within the passed list-like must match in length to the calling ``Series`` (or ``Index``),
+but ``Series`` and ``Index`` may have arbitrary length (as long as alignment is not disabled with ``join is not None``):
 
 .. ipython:: python
 
     v
-    s.str.cat([u, v], join='outer', na_rep='-')
+    s.str.cat([v, u, u.values], join='outer', na_rep='-')
```
Comment (WillAyd) [inline doc/source/text.rst]: Instead of accessing `.values` would also be preferable to use the container here, assuming you build one to instantiate a `Series` on next update

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: doc/source/text.rst [hunk 4]
```diff
@@ -306,21 +306,26 @@ The same alignment can be used when ``others`` is a ``DataFrame``:
...
+Several items can be combined a list-like container (including iterators, ``dict``-views, etc.), which may contain ``Series``, ``Index`` and ``np.ndarray``.
+Note that ``Index`` will align as well, so we change the indexes of ``s`` and ``u`` to strings for the purpose of this example:
 
 .. ipython:: python
 
-    s
-    u
-    s.str.cat([u.values,
-               u.index.astype(str).values], na_rep='-')
+    s_values = np.array(['a', 'b', 'c', 'd'], dtype=object)  # same as s.values
```
[the remark this answers is not in the document]
Comment (h-vetinari) [reply doc/source/text.rst]: Hm, I though this was clearly referencing the actual series `s` (as I wanted to motivate the variable name), rather than make a general statement about the relationship between `np.array` and `.values`.

File: doc/source/text.rst [hunk 5]
```diff
@@ -306,21 +306,26 @@ The same alignment can be used when ``others`` is a ``DataFrame``:
 Concatenating a Series and many objects into a Series
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-All one-dimensional list-likes can be combined in a list-like container (including iterators, ``dict``-views, etc.):
+Several items can be combined a list-like container (including iterators, ``dict``-views, etc.), which may contain ``Series``, ``Index`` and ``np.ndarray``.
+Note that ``Index`` will align as well, so we change the indexes of ``s`` and ``u`` to strings for the purpose of this example:
```
Comment (jreback) [inline doc/source/text.rst]: not sure what this last statement about Index actually means. can you reword.

File: doc/source/text.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (jreback) [reply doc/source/text.rst]: i agree, the comment is just confusing

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by h-vetinari** [comment]
I don't like the example too much either, especially initialising the ndarray directly. After all, what I want to do is change the index of `s` to be the same as its values (due to alignment, as I note in the running test). For this, it might make sense to wait for #22225, because then it would be very easy:

```python
>>> s2 = s.set_index(s.values)
>>> s2
a    a
b    b
c    c
d    d
dtype: object
>>> u2 = u.set_index(u.values)
>>> u2
b    b
d    d
a    a
c    c
dtype: object
>>> idx = pd.Index(['d', 'c', 'b', 'a'])
>>> idx
Index(['d', 'c', 'b', 'a'], dtype='object')
>>> u_values = u.values
>>> u_values
array(['b', 'd', 'a', 'c'], dtype=object)
>>> s2.str.cat([u2, idx, u_values], join='left')
a    aaab
b    bbbd
c    ccca
d    dddc
dtype: object
```

**Comment by jreback** [comment]
can you merge master and will look

File: doc/source/text.rst [hunk 6]
```diff
@@ -306,23 +306,30 @@ The same alignment can be used when ``others`` is a ``DataFrame``:
...
 
 .. ipython:: python
 
-    s
-    u
-    s.str.cat([u.values,
-               u.index.astype(str).values], na_rep='-')
+    s_values = np.array(['a', 'b', 'c', 'd'], dtype=object)
+    s2 = pd.Series(s_values, index=s_values)
+    s2
```
Comment (jreback) [inline doc/source/text.rst]: add this in multiple blocks as its too much to complicated here

Comment (jreback) [reply doc/source/text.rst]: this whole example needs to be simpler. maybe just leave the index as integers to avoid confusion, IOW focus less on the join in str.cat and more on the list-lke things that are going on.

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/text.rst (hunk 6 above)
Comment (h-vetinari) [reply doc/source/text.rst]: The thing this example tries to show is that it works for `Series`, `ndarray`, and `Index`, and `Index` can only be aligned with a non-integer index. This whole example would be so so much easier with #22225. I've opened an issue for the DF discussion a while ago, and asked you to comment #24046.

**Review by h-vetinari — COMMENTED** [review]
I have very limited availability over the holidays. Will consider removing the Index portion completely, but asking you to please take a look at #22225 / #24046.

**Comment by h-vetinari** [comment]
@jreback 
Incorporated your feedback (removed `Index` from example pending progress on #22225, as the extra complexity to align an index without having `Series.set_index` bloated the example).

It's all green, except for #24564

**Comment by h-vetinari** [comment]
@WillAyd 
Your review is outdated, would you care to comment (or at least remove the change request), please?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: doc/source/text.rst [hunk 7]
```diff
@@ -303,23 +303,23 @@ The same alignment can be used when ``others`` is a ``DataFrame``:
 Concatenating a Series and many objects into a Series
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-All one-dimensional list-likes can be combined in a list-like container (including iterators, ``dict``-views, etc.):
+Several items can be combined a list-like container (including iterators, ``dict``-views, etc.), which may contain ``Series``, ``Index``, ``PandasArray`` and ``np.ndarray``.
```
[the remark this answers is not in the document]
Comment (h-vetinari) [reply doc/source/text.rst]: fixed, thanks

[the remark this answers is not in the document]
Comment (h-vetinari) [reply doc/source/text.rst]: Done. I'm making an explicit list though, as only those types are allowed *within* list-likes

File: doc/source/text.rst [hunk 8]
```diff
@@ -303,23 +303,23 @@ The same alignment can be used when ``others`` is a ``DataFrame``:
...
 .. ipython:: python
 
     s
     u
-    s.str.cat([u.array,
-               u.index.astype(str).array], na_rep='-')
+    s.str.cat([u, u.array, u.to_numpy()], join='left')
 
-All elements must match in length to the calling ``Series`` (or ``Index``), except those having an index if ``join`` is not None:
+All elements without an index (e.g. ``PandasArray`` and ``np.ndarray``) within the passed list-like must match in length to the calling ``Series`` (or ``Index``),
```
[the remark this answers is not in the document]
Comment (h-vetinari) [reply doc/source/text.rst]: `PandasArray` is *very* user-facing:
```
>>> s = pd.Series(['a', 'b' ,'c', 'd'])
>>> s.array
<PandasArray>
['a', 'b', 'c', 'd']
Length: 4, dtype: object
```

and the current example was recently changed to use `.array` instead of `.values`. I think this should be documented clearly.

**Review by h-vetinari — COMMENTED** [review]
Thanks for review; pushed new commit

File: doc/source/text.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (WillAyd) [reply doc/source/text.rst]: OK thanks. I have been somewhat on the sidelines for that conversation so I'll defer to @jreback specifically on this piece

**Review by WillAyd — APPROVED** [review]
LGTM otherwise

File: doc/source/text.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (jreback) [reply doc/source/text.rst]: You don't need to mention ``PandasArray`` here, its not very interesting, nor relevant. Just say array-likes. and remove ``u.array``, the ``u.to_numpy()`` is the corrent idiom here.

[the remark this answers is not in the document]
Comment (h-vetinari) [reply doc/source/text.rst]: @jreback 
There is a big difference between `u.array` and `u.to_numpy()`:
```
>>> s = pd.Series(['a', 'b', 'c', 'd'])
>>> s.array
<PandasArray>
['a', 'b', 'c', 'd']
Length: 4, dtype: object
>>> s.to_numpy()
array(['a', 'b', 'c', 'd'], dtype=object)
```

I'm guessing `.array` will eventually replace `.values`-usage (e.g. to get rid of the index for `.str.cat`), since it is by design better suited for pandas-internal dtypes, and so the distinction above is not just an irrelevant detail IMO.

I want to show here the explicitly allowed item types to pass into a list-like, which *have* to pass:
```
nxt = others.pop(0)
[...]
if not (isinstance(nxt, (Series, Index))
        or (isinstance(nxt, np.ndarray) and nxt.ndim == 1)):
    raise ValueError(...)  # currently just a DeprecationWarning
```

Long story short, I want to show a list-like containing an `np.ndarray`, a `PandasArray` (to reiterate, this example was already changed by @TomAugspurger to use `.array` instead of `.values` in #23623), and a `Series`. (Including `Index` would be nice-to-have, but too complicated absent #22225).

[the remark this answers is not in the document]
Comment (jreback) [reply doc/source/text.rst]: @h-vetinari ``.to_numpy()`` replaces ``.values``; ``.array`` is user-accessible, but generally is not visible to by the user. its not necessary here and is just noise.

[the remark this answers is not in the document]
Comment (h-vetinari) [reply doc/source/text.rst]: @jreback 
> `.to_numpy()` replaces `.values`; 

As I did in the last few commits...

> `.array` is user-accessible, but generally is not visible to by the user. its not necessary here and is just noise.

Will remove, but pinging @TomAugspurger since he added that `u.array` in this example (although likely "just" for replacing `u.values`?).

**Review by h-vetinari — COMMENTED** [review]
PTAL

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @h-vetinari
